File: src/mobile/android/feature/settings/src/main/java/com/getsharex/xerahs/mobile/feature/settings/SettingsHubScreen.kt
/*
 * XerahS - The Avalonia UI implementation of ShareX
 * Copyright (c) 2007-2026 ShareX Team
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Optionally you can also view the license at <http://www.gnu.org/licenses/>.
 */

package com.getsharex.xerahs.mobile.feature.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Checkbox
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.getsharex.xerahs.mobile.core.data.SettingsRepository
import com.getsharex.xerahs.mobile.core.domain.selectableDestinations

@Composable
fun SettingsHubScreen(
    settingsRepository: SettingsRepository,
    onBack: () -> Unit,
    onNavigateToS3: () -> Unit,
    onNavigateToCustomUploader: () -> Unit,
    onRefresh: () -> Unit = {}
) {
    val config = settingsRepository.load()
    val s3Configured = config.s3Config.isConfigured
    val customCount = config.customUploaders.size
    var convertHeicToPng by remember { mutableStateOf(config.convertHeicToPng) }
    var selectedDestinationId by remember(config.defaultDestinationInstanceId) {
        mutableStateOf(config.defaultDestinationInstanceId)
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp)
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Button(onClick = onBack) { Text("Back") }
            OutlinedButton(onClick = onRefresh) { Text("Refresh") }
        }
        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "Settings",
            style = MaterialTheme.typography.titleLarge
        )
        Spacer(modifier = Modifier.height(16.dp))
        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
        ) {
            Card(
                modifier = Modifier.fillMaxWidth(),
                colors = CardDefaults.cardColors()
            ) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Text(
                        text = "Upload options",
                        style = MaterialTheme.typography.titleMedium
                    )
                    Spacer(modifier = Modifier.height(8.dp))
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Checkbox(
                            checked = convertHeicToPng,
                            onCheckedChange = {
                                convertHeicToPng = it
                                settingsRepository.setConvertHeicToPng(it)
                            }
                        )
                        Text(
                            text = "Convert HEIC/HEIF images to PNG before upload",
                            style = MaterialTheme.typography.bodyMedium
                        )
                    }
                    Text(
                        text = "So images display in browsers instead of prompting download.",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
            Spacer(modifier = Modifier.height(12.dp))
            Card(
                modifier = Modifier.fillMaxWidth(),
                colors = CardDefaults.cardColors()
            ) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Text(
                        text = "Active upload destination",
                        style = MaterialTheme.typography.titleMedium
                    )
                    Spacer(modifier = Modifier.height(4.dp))
                    Text(
                        text = "Choose where shared files will be uploaded. This destination is used when you share to XerahS.",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                    Spacer(modifier = Modifier.height(8.dp))
                    val options = config.selectableDestinations()
                    val effectiveId = selectedDestinationId ?: config.defaultDestinationInstanceId ?: options.firstOrNull()?.second
                    if (options.isEmpty()) {
                        Text(
                            text = "No destination configured. Set up Amazon S3 or a custom uploader below.",
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    } else {
                        options.forEach { (displayName, instanceId) ->
                            val isSelected = effectiveId == instanceId
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(vertical = 4.dp),
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                RadioButton(
                                    selected = isSelected,
                                    onClick = {
                                        selectedDestinationId = instanceId
                                        settingsRepository.setDefaultDestinationInstanceId(instanceId)
                                    }
                                )
                                Text(
                                    text = displayName,
                                    style = MaterialTheme.typography.bodyLarge,
                                    modifier = Modifier.padding(start = 8.dp)
                                )
                            }
                        }
                    }
                }
            }
            Spacer(modifier = Modifier.height(12.dp))
            Card(
                modifier = Modifier.fillMaxWidth(),
                colors = CardDefaults.cardColors()
            ) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Text(
                        text = "Upload Destinations",
                        style = MaterialTheme.typography.titleMedium
                    )
                    Spacer(modifier = Modifier.height(4.dp))
                    Text(
                        text = "Configure where your files will be uploaded. Amazon S3 and custom uploaders are supported.",
                        style = MaterialTheme.typography.bodyMedium
                    )
                }
            }
            Spacer(modifier = Modifier.height(12.dp))
            Card(
                onClick = onNavigateToS3,
                modifier = Modifier.fillMaxWidth(),
                colors = CardDefaults.cardColors()
            ) {
                Row(
                    modifier = Modifier.padding(16.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Column(modifier = Modifier.weight(1f)) {
                        Text(text = "Amazon S3", style = MaterialTheme.typography.titleSmall)
                        Text(
                            text = if (s3Configured) "Bucket: ${config.s3Config.bucketName}" else "Not configured - tap to set up",
                            style = MaterialTheme.typography.bodySmall
                        )
                    }
                }
            }
            Spacer(modifier = Modifier.height(8.dp))
            Card(
                onClick = onNavigateToCustomUploader,
                modifier = Modifier.fillMaxWidth(),
                colors = CardDefaults.cardColors()
            ) {
                Row(
                    modifier = Modifier.padding(16.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Column(modifier = Modifier.weight(1f)) {
                        Text(text = "Custom Uploader", style = MaterialTheme.typography.titleSmall)
                        Text(
                            text = if (customCount > 0) "$customCount uploader(s)" else "Not configured - tap to add",
                            style = MaterialTheme.typography.bodySmall
                        )
                    }
                }
            }
        }
    }
}

File: src/mobile/ios/XerahSMobile/Features/Settings/SettingsHubScreen.swift
//
//  SettingsHubScreen.swift
//  XerahS Mobile (Swift)
//
//  XerahS - The Avalonia UI implementation of ShareX
//  Copyright (c) 2007-2026 ShareX Team
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
//  Optionally you can also view the license at <http://www.gnu.org/licenses/>.
//

import SwiftUI

struct SettingsHubScreen: View {
    let settingsRepository: SettingsRepository
    var onBack: () -> Void
    var onNavigateToS3: () -> Void
    var onNavigateToCustomUploader: () -> Void

    @State private var convertHeicToPng: Bool = true

    private var config: ApplicationConfig { settingsRepository.load() }

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            HStack {
                Button("Back", action: onBack)
                Spacer()
            }
            .padding(.horizontal)

            Text("Settings")
                .font(.title2)
                .padding(.horizontal)

            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    VStack(alignment: .leading, spacing: 4) {
                        Text("Upload options")
                            .font(.headline)
                        Text("Convert HEIC/HEIF images to PNG before upload so they display in browsers instead of prompting download.")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(16)
                    .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 10))

                    Toggle(isOn: Binding(
                        get: { convertHeicToPng },
                        set: { newValue in
                            convertHeicToPng = newValue
                            settingsRepository.setConvertHeicToPng(newValue)
                        }
                    )) {
                        Text("Convert HEIC/HEIF to PNG before upload")
                            .font(.subheadline)
                    }
                    .padding(16)
                    .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 10))
                    .onAppear { convertHeicToPng = settingsRepository.getConvertHeicToPng() }

                    VStack(alignment: .leading, spacing: 4) {
                        Text("Upload Destinations")
                            .font(.headline)
                        Text("Configure where your files will be uploaded. Amazon S3 and custom uploaders are supported.")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(16)
                    .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 10))

                    Button(action: onNavigateToS3) {
                        HStack {
                            VStack(alignment: .leading, spacing: 2) {
                                Text("Amazon S3")
                                    .font(.subheadline.weight(.medium))
                                Text(config.s3Config.isConfigured ? "Bucket: \(config.s3Config.bucketName)" : "Not configured - tap to set up")
                                    .font(.caption)
                                    .foregroundStyle(.secondary)
                            }
                            Spacer()
                        }
                        .padding(16)
                        .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 10))
                    }
                    .buttonStyle(.plain)

                    Button(action: onNavigateToCustomUploader) {
                        HStack {
                            VStack(alignment: .leading, spacing: 2) {
                                Text("Custom Uploader")
                                    .font(.subheadline.weight(.medium))
                                Text(config.customUploaders.isEmpty ? "Not configured - tap to add" : "\(config.customUploaders.count) uploader(s)")
                                    .font(.caption)
                                    .foregroundStyle(.secondary)
                            }
                            Spacer()
                        }
                        .padding(16)
                        .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 10))
                    }
                    .buttonStyle(.plain)
                }
                .padding(.horizontal)
            }
        }
    }
}
